automatilib/cola_auth/i_dot_ai_cola_views.py

```diff
@@ -0,0 +1,149 @@
+        response = redirect(reverse(redirect_url or "index"))
+        response.delete_cookie(settings.COLA_COOKIE_NAME)
+        return response
+
+
+class IAIColaLogin(MethodDispatcher):
+    @abstractmethod
+    def pre_login(self):
+        """
+        A method that is invoked after checking if a user is already logged in,
+        but before performing any actions on the request
+        """
+        pass
+
+    @abstractmethod
+    def post_login(self):
+        """
+        A method that is invoked after `handle_claims` and logging/authenticating a user,
+        but before returning an HTTP response
+        """
+        pass
+
+    @abstractmethod
+    def handle_claims(self, user: User, token_payload: dict) -> None:
+        """
+        A method that is invoked after logging/authenticating a user and before `post_login`,
+        but before returning an HTTP response.
+        Email, family name and given name are already taken from the token and saved to the user object before this
+        :param user: The user that is logged in and authenticated
+        :param token_payload: The token payload that includes information from COLA
+        """
+        pass
+
+    def get(self, request: HttpRequest) -> HttpResponse:
+        redirect_url = settings.LOGIN_REDIRECT_URL
+        if request.user.id:
+            return redirect(reverse(redirect_url or "index"))
+
+        self.pre_login()
+
+        if not (cola_cookie := request.COOKIES.get(settings.COLA_COOKIE_NAME, None)):
+            LOGGER.error("No cookie found")
+            return HttpResponseServerError()
+        if not (
+            regex_match := re.search(
+                settings.COLA_JWT_EXTRACTION_REGEX_PATTERN,
+                unquote(cola_cookie),
+            )
+        ):
+            LOGGER.error("No cookie regex match found")
+            return HttpResponseServerError()
+
+        cola_cognito_user_pool_jwk, cola_issuer = get_cola_cognito_user_pool_jwk()
+        header = jwt.get_unverified_header(regex_match.group(0))
+        jwt_kid = header["kid"]
+        public_key = next(key for key in cola_cognito_user_pool_jwk["keys"] if key["kid"] == jwt_kid)
+
+        try:
+            token = regex_match.group(0)
+            token = ".".join(token.split(".")[:3])
+            payload = jwt.decode(
+                token=token,
+                audience=settings.COLA_COGNITO_CLIENT_ID,
+                issuer=cola_issuer,
+                algorithms=["RS256"],
+                key=public_key,
+                options={
+                    "require_aud": True,
+                    "require_iat": True,
+                    "require_exp": True,
+                    "require_iss": True,
+                    "require_sub": True,
+                },
+            )
+
+        except (ExpiredSignatureError, JWTClaimsError, JWTError, KeyError) as error:
+            LOGGER.error("cookie error:", type(error).__name__)
+            return HttpResponseServerError()
```

Comment:
all (?) of the HttpResponseServerError are actually 40X client errors, e.g. token expired etc?

Reply:
I think keyerror might be a 500 error but the rest I agree are probably 400, any idea which 4xx each is?